### PR TITLE
Support secrets integration in wasmCloud

### PR DESCRIPTION
**Pull request by brooksmtownsend**
## Feature or Problem
This PR implements wasmCloud support for fetching and storing secrets securely for wasmCloud applications on behalf of components and capability providers. Capability providers and components can have secrets passed to them as a part of their startup config, and capability providers can receive secrets passed to them as a source or target config via a link.

**Secrets in wasmCloud are treated as a special piece of configuration.** Secrets are actually stored in encrypted and secure stores, all wasmCloud stores are _secret references_ which are pieces of config that start with `secret_` and have the following structure:
```rust
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
struct SecretReference {
    /// The backend to use for retrieving the secret.
    pub backend: String,
    /// The key to use for retrieving the secret from the backend.
    pub key: String,
    /// The version of the secret to retrieve. If not supplied, the latest version will be used.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version: Option<String>,
}
```

So, a NATS KV secret represented in JSON could be named `secret_my-password`:
```json
{
"backend": "nats-kv",
"key": "my-password",
// optional version
}
```

## Related Issues
#2190 
As noted in #2389, this PR will _not_ include support for sharing the secret resource using a custom interface between components or providers. A limitation of secrets support in wasmCloud for the initial release is that if a secret value is revealed by a component and transmit over wRPC, it will be transmit in plaintext (over a hopefully TLS secured NATS connection) so it's critical that developers reveal secrets only in time to use them.
See #2395 for an example of this in action

## Release Information
wasmCloud 1.1
next minor wascap
next minor wasmcloud-core
next minor provider-archive
next minor provider-sdk

#### Planning / WIP / tasklist

Starting with the config integration test, I'm planning on:

- [x] Create/modify a test that I can use for quick iteration
- [x] Expose the wasmCloud secrets topic parameter as a command line flag and host config
- [x] Ensure component can fetch a secret through the plumbing, and reveal it
- [x] Extend plumbing to "fetch" a secret value from a regular `nats reply` utility
- [x] Integrate with the NATS KV secrets backend
- [x] Modify original test to launch the NATS KV secrets backend for full integration
- [x] Clean up TODOs around the code for efficiency, open questions, etc
- [x] Support secrets for components
- [x] Support secrets for providers
- [x] Support delivering secret link values to providers upon startup (note: we don't support source/target config for components today, so we can add secrets there when we add it. https://github.com/wasmCloud/wasmCloud/issues/2170)

To be done in a followup:
- [ ] Support delivering secret link values to providers over the lattice using a shared encryption xkey (detailed in #2407)
- [ ] Test edge cases and ensure developer experience for missing / unauthenticated / wrong secrets is good enough to self-service (I've done some testing here around the experience, but more is needed especially around error messages)

Considerations
- [x] Should we be returning options if secrets don't exist? ~(I think so, rather than a concrete value)~ Discussed with @protochron offline, we should build in an error type that can be used with `get` but not with `reveal`
- [x] Should we should be fetching the secret in context of who's revealing it, or who owns the resource? We should be revealing it in the context of who owns the resource, but we need clarity on how wRPC will work here.

### CODE REVIEW

**Review by vados-cosmonic — COMMENTED** [review]
Looks great -- Mostly nits!

File: crates/core/src/secrets.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+use std::fmt::{Debug, Formatter, Result};
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Clone)]
+// This tagging allows deserializers to know whether the secret is a string or bytes.
+// This is especially necessary for languages where strings and bytes are treated very similarly.
+#[serde(tag = "kind", content = "value")]
+pub enum SecretValue {
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/core/src/secrets.rs:9]: I don't think there's a risk here, and honestly the main use of this enum is to serialize it directly and push it to the providers so we'll take on the risk of a new tagged value coming in just in our SDKs

File: crates/host/src/secrets.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    pub version: Option<String>,
+}
+
+#[derive(Debug)]
+/// A manager for fetching secrets from a secret store, caching secrets clients for efficiency.
+pub struct Manager {
+    config_store: Store,
+    secret_store_topic: Option<String>,
+    nats_client: Client,
+    backend_clients: Arc<RwLock<HashMap<String, Arc<WasmcloudSecretsClient>>>>,
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:34]: Left a comment!
```
    /// A map of backend names, e.g. nats-kv or vault, to secrets clients, used to cache clients for efficiency.
```

File: crates/host/src/secrets.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    /// The version of the secret to retrieve. If not supplied, the latest version will be used.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub version: Option<String>,
+}
+
+#[derive(Debug)]
+/// A manager for fetching secrets from a secret store, caching secrets clients for efficiency.
+pub struct Manager {
+    config_store: Store,
+    secret_store_topic: Option<String>,
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:32]: Left a comment!
```
    /// The topic to use for configuring clients to fetch secrets from the secret store.
```

File: crates/host/src/secrets.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+                "secret store not configured, could not create secrets client"
+            ));
+        };
+
+        // If we already have a client for this backend, return it
+        if let Some(client) = self.backend_clients.read().await.get(backend) {
+            return Ok(client.clone());
+        }
+
+        // NOTE(brooksmtownsend): This isn't an `else` clause to ensure we drop the read lock
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:74]: I'm a fan of doing this for clarity 👍🏻

File: crates/host/src/secrets.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    nats_client: Client,
+    backend_clients: Arc<RwLock<HashMap<String, Arc<WasmcloudSecretsClient>>>>,
+}
+
+impl Manager {
+    /// Create a new secret manager with the given configuration store, secret store topic, and NATS client.
+    ///
+    /// All secret references will be fetched from this configuration store and the actual secrets will be
+    /// fetched by sending requests to the configured topic. If the provided secret_store_topic is None, this manager
+    /// will always return an error if [`Self::fetch_secrets`] is called with a list of secrets.
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/host/src/secrets.rs:42]: Hmnn if the goal is making sure that the host always has a secrets manager it can call, can we solve that at the host level? It feels like `SecretsManager` shouldn't exist if the topic isn't there...

How do you feel about implementing trait stuff on `Option<SecretsManager>`? that way the host could have it but the ergonomics wouldn't be so bad?

File: src/main.rs [hunk 6]
```diff
@@ -177,6 +177,10 @@ struct Args {
     )]
     policy_timeout_ms: Option<Duration>,
 
+    /// If provided, enables interfacing with a secrets backend for secret retrieval
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply src/main.rs:180]: It just matches our current precedent for the policy service topic which was the main reason I went in this direction, I think having a default topic prefix would be fine the main case I want to avoid is passing a topic prefix and forgetting to enable the support

File: tests/common/secrets.rs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            wasmcloud_secrets_types::WASMCLOUD_HOST_XKEY,
+            request_xkey
+                .public_key()
+                .parse::<async_nats::HeaderValue>()
+                .unwrap(),
+        );
+
+        // NOTE: This is just demonstrative that you only need the public key to seal the secret
+        let transit_xkey_pub = nkeys::XKey::from_public_key(&self.transit_xkey.public_key())
+            .expect("public key to be valid");
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply tests/common/secrets.rs:58]: I mean, it could basically be a safety comment given we just generated it, but in a test I wasn't really worried about this panic-ing

File: tests/common/secrets.rs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+        Ok(())
+    }
+
+    pub async fn start(&self) -> Result<BackgroundServer> {
+        let res = BackgroundServer::spawn(
+            tokio::process::Command::new(
+                std::env::var("WASMCLOUD_NATS_KV_SECRETS_BACKEND")
+                    .as_deref()
+                    .unwrap_or("./target/debug/secrets-nats-kv"),
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply tests/common/secrets.rs:93]: Is there a better way to do that? Is that like the `CARGO_ROOT_DIR` env var? I can take a look but if you had it offhand

File: tests/common/secrets.rs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            )
+            .args([
+                "--encryption-xkey-seed",
+                &self.encryption_xkey.seed().expect("seed to be valid"),
+                "--transit-xkey-seed",
+                &self.transit_xkey.seed().expect("seed to be valid"),
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply tests/common/secrets.rs:99]: I figure this is fine given this is a test and we control the generation of the keypair (this only fails if we constructed the xkey from a public key)

File: crates/secrets-client/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,176 @@
+use async_nats::HeaderMap;
+use nkeys::XKey;
+use wasmcloud_secrets_types::{
+    Secret, SecretRequest, SecretResponse, RESPONSE_XKEY, WASMCLOUD_HOST_XKEY,
+};
+
+const DEFAULT_API_VERSION: &str = "v0";
+
+#[derive(Debug)]
```
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply crates/secrets-client/src/lib.rs:9]: oh yeah that's totally fine! I thought maybe you were trying to avoid a dep or something, this definitely isn't necessary!

File: src/main.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (vados-cosmonic) [reply src/main.rs]: Ah OK that's pretty reasonable -- having  the one can certainly (and does) work with policy service

File: crates/host/src/secrets.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:44]: I don't mind that option, mild concern about ease of tracking down bugs with traits on an option, but I still do prefer the current method of a short-circuit just like we do for policy requests. I like keeping the pattern we have in place for now but let me know if you feel strongly about it

**Comment by brooksmtownsend** [comment]
So many conflicts 🫣

File: crates/core/src/secrets.rs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    String(String),
```
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/core/src/secrets.rs:10]: I did see further down that this type implements some secrecy stuff so maybe this doesn't matter?

File: crates/secrets-client/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+pub enum SecretClientError {
```
Comment (thomastaylor312) [inline crates/secrets-client/src/lib.rs:10]: Super Nit: I'd just recommend using `thiserror` rather than manually implementing display

File: crates/secrets-client/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+#[derive(Debug)]
+pub struct Client {
+    client: async_nats::Client,
+    topic: SecretsTopic,
+    server_xkey: XKey,
+}
+
+impl Client {
+    pub async fn new(
```
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/secrets-client/src/lib.rs:77]: Since this is going to be used in the NATS impl and probably the community provided vault one, we probably should have this documented before we merge

File: crates/secrets-client/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+        backend: &str,
+        prefix: &str,
+        nats_client: async_nats::Client,
+    ) -> Result<Self, SecretClientError> {
```
Comment (thomastaylor312) [inline crates/secrets-client/src/lib.rs:81]: Nit: Suggest a pub type alias for this result `Result<T, SecretClientError>`

File: crates/secrets-client/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+                SecretClientError::RequestServerXkeyError => "Request Server Xkey Error",
+                SecretClientError::InvalidXkeyError => "Invalid Xkey Error",
+                SecretClientError::SealSecretRequestError => "Seal SecretRequest Error",
+                SecretClientError::SendSecretRequestError => "Send SecretRequest Error",
+                SecretClientError::SerializeSecretRequestError => "Serialize SecretRequest Error",
+                SecretClientError::ParseServerResponseXkeyError => "Parse Server Response Error",
+                SecretClientError::OpenSecretResponseError => "Open SecretResponse Error",
+                SecretClientError::DeserializeSecretResponseError =>
+                    "Deserialize SecretResponse Error",
+                SecretClientError::ServerError(_) => "Server Error",
```
Comment (thomastaylor312) [inline crates/secrets-client/src/lib.rs:42]: Should we write the inner server error message out for a clearer error?

File: tests/common/secrets.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply tests/common/secrets.rs:58]: Generally my rule of thumb is if a test function returns a Result to not panic, otherwise, panicking is fine. This is more of a nit for me though in this specific case

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Mostly minor things and some nits

File: crates/core/src/secrets.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/core/src/secrets.rs:10]: There are a few different enums/structs with the name `SecretValue`, so perhaps it would be best to rename some of these for clarity. I don't implement the secrecy crate here since I didn't want the representation to change and be complicated to work with from, say, the Go provider SDK, but maybe if I implement serialize it wouldn't matter. 

If we're cool with `core` depending on the secrecy crate, I can give it a shot. Slightly concerned with the tagged enum piece of it bit I'm sure I could get it working

File: crates/host/src/secrets.rs [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+        );
+
+        // If we don't have an entity JWT, we can't provide its identity to the secrets backend
+        let Some(entity_jwt) = entity_jwt else {
+            bail!(
+                "entity did not have an embedded JWT, required to fetch secrets (was this entity signed during build?)"
+            );
+        };
+
+        let secrets = stream::iter(secret_names.iter())
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:135]: I can, with an extra borrow. What exactly does this change in terms of efficiency?

File: crates/host/src/secrets.rs [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+                    name: secret_ref.key.clone(),
+                    version: secret_ref.version.clone(),
+                    context: Context {
+                        entity_jwt: entity_jwt.to_string(),
+                        host_jwt: host_jwt.to_string(),
+                        application: application.cloned().map(|name| Application { name }),
+                    },
+                };
+                secrets_client
+                    .get(request, nkeys::XKey::new())
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:162]: Yeah that's the idea, we mint an XKey for secret request so that requests can't be replayed with the same payload

File: crates/host/src/secrets.rs [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+                    .await
+                    .map_err(|e| anyhow::anyhow!(e))
+            })
+            // Build the map of secrets depending on if the secret is a string or bytes
+            .try_fold(HashMap::new(), |mut secrets, secret_result| async move {
+                match secret_result {
+                    WasmcloudSecret {
+                        name,
+                        string_secret: Some(string_secret),
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/secrets.rs:171]: When setting this up I was considering that as an invalid case, either your secret is binary or a string, so I suppose here in this case if it was both then you would get that string secret. Do you think the other case should be an error, or just a warning?

File: crates/host/src/wasmbus/handler.rs [hunk 19]
```diff
@@ -273,3 +280,36 @@ impl Logging for Handler {
...
+
+    async fn reveal(
+        &self,
+        secret: secrets::store::Secret,
+    ) -> anyhow::Result<secrets::store::SecretValue> {
+        let read_lock = self.secrets.read().await;
+        let Some(secret_val) = read_lock.get(secret.as_str()) else {
+            // NOTE(brooksmtownsend): This error case should never happen, since we check for existence during `get` and
+            // fail to start the component if the secret is missing. We might hit this during wRPC testing with resources.
+            let error_msg = "secret not found to reveal, ensure the secret is declared and associated with this component at startup";
```
[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/host/src/wasmbus/handler.rs:306]: Sure 👍🏻

File: crates/secrets-client/src/lib.rs (hunk 12 above)
Comment (brooksmtownsend) [reply crates/secrets-client/src/lib.rs:10]: I was hoping to hold the nits for the `secrets-client` crate for a followup since there are a couple of things we'd like to improve, like Victor's comments that are captured in #2411 . Do you have an aversion to that?

File: crates/host/src/secrets.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/host/src/secrets.rs]: Not too much honestly. Just observation

File: crates/host/src/secrets.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/host/src/secrets.rs:162]: I looked it up and it costs 13 nanoseconds in benchmarks. I think we're fine 😅

File: crates/secrets-client/src/lib.rs (hunk 12 above)
Comment (thomastaylor312) [reply crates/secrets-client/src/lib.rs:10]: Nope that is fine

**Review by thomastaylor312 — APPROVED** [review]
